### ISSUE TITLE
[Feature Request] Hardware Encoding in Unraid

### ISSUE THREAD

**Issue by dslowder** [Josh5/unmanic#28]
This was mentioned in [Issue #20](https://github.com/Josh5/unmanic/issues/20), but I didn't see the separate FR tracked that you mentioned creating there.

I think the current release is extremely usable, I've had it chugging along for weeks now. But being able to speed things along with hardware encoding would be amazing. At my current rate, it'll be 3+ years before everything in x265.

**Comment by Josh5** [comment]
This is on my todo list for sure. At the moment tho I wanted to lock down all the main configuration features of the application and get it tidied up to a point where we can have a initial release.
I'm guessing that this may be about a month or so away.
Once that is done, HW accelerated encoding should come next.

Thanks for using the application. And thanks for expressing interest in moulding it's future

**Comment by m0ngr31** [comment]
I was under the impression that CPU encoding produces better quality than GPU encoding?

**Comment by dslowder** [comment]
Anytime you transcode a file, there is going to be some quality loss, even using CPU encoding. There is supposed to be a larger loss in quality when using hardware encoding (GPU or even something like Intel Quicksync). How noticeable that quality loss is depends on a lot of factors.

Hardware transcoding is significantly faster though. For my setup, being able to use hardware encoding on my P2000 would be orders of magnitude faster. I have an older CPU and I'm not dedicating all cores to unmanic so I only get about 1 frame every 5 seconds. To me, the minor quality difference isn't worth the additional time.

**Comment by nphil** [comment]
Please consider prioritizing this, I have an intel iGPU that's just sitting idle most of the time and using Quicksync to hardware encode would be awesome!

**Comment by adznon** [comment]
Hi Josh5

Am loving your application, would like to put my hand up for GPU hardware encoding like @dslowder I also have a P2000 in my UNRaid setup that i would like to put to work using unmanic rather than just for Plex

Any expected eta on this ?

**Comment by Josh5** [comment]
Glad to hear you are enjoying the application.

This issue is dependent on me finishing and merging in the changes in the "dev-settings" branch. Those changes are needed in order to have the ability for more than one encoder for a particular codec as is required for HW encoding.
It is getting close to being done, but I also need to tidy up the history logging as that is causing some issues also.
I cannot give any kind of eta as all development is done dependent on my spare time and I am unsure exactly how much of that I have in the future. But this is definitely a priority feature that I'm working towards.

**Comment by adznon** [comment]
Hello Josh,

Hope your well.

Have you had any free time to look at this yet ?

Thanks
Adam

**Comment by maximberezin97** [comment]
[Looks like Josh5 is taking a break from development until February/March 2020.](https://forums.unraid.net/topic/76878-support-josh5-unmanic-library-optimiser/?do=findComment&comment=796074)

I don't know much about hardware transcoding, but if there are any developers out there with experience working on other hardware-accelerated applications in the unRAID environment (Plex, Emby, Jellyfin, AI frameworks like TensorFlow or PyTorch) and specifically the Linuxserver.io unRAID NVIDIA build plugin, their skills would be very useful here!

**Comment by jfroebe** [comment]
If the container is started with "--gpus all" or similar the following nvidia devices will appear within the container:

/dev/nvidia0  /dev/nvidiactl

if they exist, ffmpeg (assuming it is compiled with hw support.  parse ffmpeg -hwaccels output ) can be called with the standard hw acceleration params.  If the hardware encode fails, unraid should attempt a non-hw accel encode.

**Comment by jfroebe** [comment]
I looked into it a bit more.  For Alpine Linux the cuda toolkit and the ffmpeg binaries will have to be built from source.  The cuda toolkit isn't available in the standard repository and the ffmpeg build in the repository wasn't compiled with the nvidia codecs.

Another container was able to get the cuda toolkit going with Alpine: https://github.com/clear-datacenter/plan/wiki/alpine--cuda

Same for custom building ffmpeg:  https://github.com/denismakogon/ffmpeg-alpine/blob/master/build-stage/Dockerfileents  

Pre docker container build requirements:  install nvidia driver, nvidia-container-toolkit.  see https://github.com/NVIDIA/nvidia-docker for more details

**Comment by Josh5** [comment]
NVIDIA HW encoding is complete
